demos/qml_optimization.ipynb

```diff
@@ -0,0 +1,469 @@
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 41,
+      "metadata": {
+        "id": "nZjOeD-79zKT"
+      },
+      "outputs": [],
+      "source": [
+        "params = {\"weights\": weights, \"bias\": bias}\n",
+        "\n",
+        "@qml.qjit\n",
+        "def optimization(params, data, targets):\n",
+        "    opt_state = opt.init(params)\n",
+        "    args = (params, opt_state, data, targets)\n",
+        "    (params, opt_state, _, _) = catalyst.for_loop(0, 100, 1)(update_step)(args)\n",
+        "    return params"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "dsFXWti89HQ_"
+      },
+      "source": [
+        "Note that we use `catalyst.for_loop` rather than a standard Python for loop, to allow the control flow to be JIT compatible."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 42,
+      "metadata": {
+        "id": "gVRxMBgg2nGF"
+      },
+      "outputs": [],
+      "source": [
+        "final_params = optimization(params, data, targets)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 43,
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "kSJlW1rx2tyR",
+        "outputId": "a5020f22-fcc8-4921-93cb-81fea1a711f2"
+      },
+      "outputs": [
+        {
+          "name": "stdout",
+          "output_type": "stream",
+          "text": [
+            "{'bias': array(-0.75292885), 'weights': array([[ 1.63086995,  1.55018972,  0.6721261 ],\n",
+            "       [ 0.7266062 ,  0.36422543, -0.756247  ],\n",
+            "       [ 2.78387487,  0.62721014,  3.44996393],\n",
+            "       [-1.10119515, -0.12679488,  0.89283774],\n",
+            "       [ 1.27236329,  1.10631134,  2.22051434]])}\n"
+          ]
+        }
+      ],
+      "source": [
+        "print(final_params)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## Timing the optimization"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "m2I1r0Bm9mzp"
+      },
+      "source": [
+        "We can time the two approaches (JIT compiling just the cost function, vs JIT compiling the entire optimization loop) to explore the differences in performance:"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 46,
+      "metadata": {
+        "id": "I9Yaeyer9rVb"
+      },
+      "outputs": [],
+      "source": [
+        "def optimization_noqjit(params):\n",
+        "    opt = optax.adam(learning_rate=0.3)\n",
```

Comment:
@josh146 you may want to re-run this after removing the `optax.adam` construct from this function for a fair comparison.